morf-core/src/main/java/org/alfasoftware/morf/jdbc/ResultSetComparer.java

```diff
@@ -707,44 +673,59 @@ public interface CompareCallback {
+             if (!leftRs.isBeforeFirst() || !rightRs.isBeforeFirst()) {
+                throw new IllegalStateException(format("The following queries should return at least one record: [%s]%n[%s]", leftRs.getStatement(), leftRs.getStatement()));
              }
           } catch (SQLException e) {
             ExceptionUtils.rethrow(e);
           }
         }
      },
 
-     NON_ZERO_RECORD_COUNT_ON_RIGHT() {
-        @Override
-        void validate(ResultSet leftRs, ResultSet rightRs) {
-           try {
-              if (!rightRs.isBeforeFirst()) {
-                 throw new IllegalStateException(format("The following query should return at least one record: [%s]", rightRs.getStatement()));
-              }
-           } catch (SQLException e) {
-             ExceptionUtils.rethrow(e);
-           }
-        }
-     };
-
-     /**
-      * Handle the validation.
-      *
-      * @param leftRs the left {@link ResultSet} to validate.
-      * @param rightRs the right {@link ResultSet} to validate.
-      */
-     abstract void validate(ResultSet leftRs, ResultSet rightRs);
+    /**
+     * Validates that the left and right result values are both numeric and are not equal to zero.
+     *
+     * @author Copyright (c) Alfa Financial Software Limited. 2024
+     */
+    NON_ZERO_RESULT {
+       @Override
+       void validateSingleResult(Comparable<?> leftValue, Comparable<?> rightValue) {
+         if (!(leftValue instanceof BigDecimal) || !(rightValue instanceof BigDecimal)) {
+           throw new IllegalStateException(format("Incorrect value type for this validator. Expecting a BigDecimal but got [%s] for the left value and [%s] for the right value.", leftValue.getClass().getName(), rightValue.getClass().getName()));
+         }
+
+         if (((BigDecimal) leftValue).equals(BigDecimal.ZERO) && ((BigDecimal) rightValue).equals(BigDecimal.ZERO)) {
+           throw new IllegalStateException("Left and right record queries returned zero");
+         }
+      }
+    };
+
+    /**
+     * Handles validation of the left and right {@link ResultSet} prior to the first invocation of {@code ResultSet#next()}.
+     *
+     * @param leftRs the left {@link ResultSet} to validate.
+     * @param rightRs the right {@link ResultSet} to validate.
+     */
+    @SuppressWarnings("unused")
+    void validateBeforeNext(ResultSet leftRs, ResultSet rightRs) { /* Default empty implementation */ }
+
+    /**
+     * Handles validation of the left and right {@link Comparable} values when a single result is expected.
+     *
+     * @param leftValue the left {@link Comparable} value to validate.
+     * @param rightValue the right {@link Comparable} value to validate.
+     */
+    @SuppressWarnings("unused")
+    void validateSingleResult(Comparable<?> leftValue, Comparable<?> rightValue) { /* Default empty implementation */ }
   }
```

Comment:
Name of the `validateBeforeNext()` method feels somewhat confusing to me, because it is not being executed before each "next", which is how I seem to be understanding that name.

Perhaps `validateBeforeFirstRow()` or `validateBeforeComparison()` would be more self-documenting?

All this said, this is an internal part of this class, and thus is not very visible to the outside world, apart from the stacktrace of the validation exception.